chain/network/src/cache.rs

```diff
@@ -0,0 +1,398 @@
+/// that require route back being dropped.
+///
+/// Solution:
+///
+/// The cache will accept new elements until it is full. If it receives a new
+/// element but it implies going over the capacity, the message to remove is
+/// selected as following:
+///
+/// 1. For every message store the time it arrives.
+/// 2. For every peer store how many message should be routed to it.
+///
+/// First are removed messages that have been in the cache more time than
+/// EVICTED_TIMEOUT. If no message was removed, it is removed the oldest
+/// message from the peer with more messages in the cache.
+///
+/// Rationale:
+///
+/// - Old entries in the cache will be eventually removed (no memory leak).
+/// - If the cache is not at full capacity, all new records will be stored.
+/// - If a peer try to abuse the system, it will be able to allocate at most
+///     $capacity / number_of_active_connections$ entries.
+pub struct RouteBackCache {
+    /// Maximum number of records allowed in the cache.
+    capacity: u64,
+    /// Maximum time allowed before removing a record from the cache.
+    evict_timeout: u64,
+    /// Main map from message hash to time where it was created + target peer
+    /// Size: O(capacity)
+    main: HashMap<CryptoHash, (Time, PeerId)>,
+    /// Number of records allocated by each PeerId.
+    /// The size is stored with negative sign, to order in PeerId in decreasing order.
+    /// Size: O(number of active connections)
+    size_per_target: BTreeSet<(Size, PeerId)>,
+    /// List of all hashes associated with each PeerId. Hashes within each PeerId
+    /// are sorted by the time they arrived from older to newer.
+    /// Size: O(capacity)
+    record_per_target: BTreeMap<PeerId, BTreeSet<(Time, CryptoHash)>>,
+}
+
+impl RouteBackCache {
+    fn is_full(&self) -> bool {
+        self.capacity == self.main.len() as u64
+    }
+
+    fn remove_frequent(&mut self) {
+        let (mut size, target) = self.size_per_target.iter().next().cloned().unwrap();
+
+        if let btree_map::Entry::Occupied(mut entry) = self.record_per_target.entry(target.clone())
+        {
+            {
+                let records = entry.get_mut();
+                let first = records.iter().next().cloned().unwrap();
+                records.remove(&first);
+                self.main.remove(&first.1);
+            }
+
+            if entry.get().is_empty() {
+                entry.remove();
+            }
+        }
+
+        self.size_per_target.remove(&(size, target.clone()));
+        // Since size has negative sign, adding 1, is equivalent to subtracting 1 from the size.
+        size += 1;
+
+        if size != 0 {
+            self.size_per_target.insert((size, target));
+        }
+    }
+
+    fn remove_evicted(&mut self) {
+        if self.is_full() {
+            let now = Utc::now().timestamp_millis() as Time;
+            let remove_until = now.saturating_sub(self.evict_timeout);
+
+            let mut new_size_per_target = BTreeSet::new();
```

Comment:
Yes I got confused, for some reason thought that since I was holding a mutable reference from an object contained in self, I couldn't mutate another field in self, but it is allowed which makes sense.